Comments Redesign: Always expanded view

Copons opened the pull request:
Fix #19605
Fix #19306 

Alternative take to #19472
All props to @vindl and @drw158 for bringing this forward!

Experimental branch to test the alternative approach with all comments expanded.

Also applied `pending` style to expanded view to make it look the same as collapsed.

| Big Screens | Small Screens | Bulk Mode |
| --- | --- | --- |
| <img width="770" alt="screen shot 2017-11-13 at 14 40 36" src="https://user-images.githubusercontent.com/2070010/32731113-e22e931a-c880-11e7-9482-7d579a079b48.png"> | <img width="473" alt="screen shot 2017-11-13 at 14 41 24" src="https://user-images.githubusercontent.com/2070010/32731116-e6afbc2a-c880-11e7-9419-83aafe705903.png"> | <img width="744" alt="screen shot 2017-11-13 at 14 42 15" src="https://user-images.githubusercontent.com/2070010/32731121-eac8cc70-c880-11e7-9ab5-bc05d61bd720.png"> |

### Testing instructions

Checkout this branch and run it with
`env ENABLE_FEATURES=comments/management/m3-design npm start`

drw158 commented:
Liking it!

The only major blocker I see is the reply field. We'd need to add a reply button to the actions. But that would total 6, including the edit button. That would be pretty cramped... but I think it would work.

Copons commented:
@drw158 I'd actually urge you to test it on a real site, if you didn't already.
You'd notice how it's not entirely clear when a comment ends and the other begins.

Options:
- Restore the top/bottom margins (which also gave the little expand animation).
- Drop the inner borders (bottom of header, top of reply, top of actions on small screens).

About the reply action: is it needed, considering the reply fields is always visible?
Or, and I may have missed/forgot about it, there was a decision to hide the reply area by default and only show it when clicking the Reply button?

gwwar commented:
I do like seeing all of the comment content! 

I agree with the need for some margin/padding between each comment.

Some ideas:
- Do not display the quick reply field
- Add the reply button as an action. 
-- For now clicking on it can add the field enabled
-- In the future it might link to the post-specific view with that comment highlighted. It feels like it'd be good to have some extra context when replying.

drw158 commented:
Hm, good point, it's hard to scan the comments without the reply field.

 I'd like to find a solution that doesn't add margins below the comments, but it certainly helps readability. For mobile, it would be ideal to not have margins, so maybe if we solve the problem on mobile it will carry over to desktop. Here is an idea, using the M3 i3 design:

![screen shot 2017-11-07 at 2 17 30 pm](https://user-images.githubusercontent.com/618551/32515671-794ce0ea-c3c6-11e7-83d3-8026aa15e3bb.png)

Copons commented:
Looking good @drw158! I'm going to try it asap.

Meanwhile, I've got two other nitpicks:

- The post title: the more I look at it, the more it's not clear it's the post title when the comment is expanded.
In collapsed view it's near the author name, e.g. `Author Name > Post Title`, which makes sense to me, and it gives an idea of the "flow" of the conversation (that `>` chevron makes me think of "on", as in "This user (commented) on that post").
In expanded instead, it could easily be the first paragraph of the comment, in bold weight.
Would you be opposed if we kept the same style and position the post title has in collapsed view, in expanded view as well?

- The date!
The idea was to have a relative date (`2 days ago`) in collapsed view, and the full date (a localized version of `6 November 2017, 10:50`) in expanded.
If we basically drop the collapsed view, we're also basically dropping the relative date, which was a really nice addition imho.
What about we show the relative date by default with a tooltip with the full date? Would it be too bad a11y-wise?

drw158 commented:
> Would you be opposed if we kept the same style and position the post title has in collapsed view, in expanded view as well?

Not opposed, but I'm not sure of the history of why the post title was moved underneath. I do like it underneath...was the reason to make room for the user info button?

I have been thinking of ways to make the post title stand out:

- border around the title to make it look like an 'attachment'
- icon next to title 

> The date!

I think we can use the full date on the comment specific view, or the edit comment view. Relative date seems fine for this view.

vindl commented:
> I agree with the need for some margin/padding between each comment.

+1. This was one of the main motivations behind my previous PR with the aim of improving overall readability.

rodrigoi commented on client/my-sites/comments/comment/index.jsx:
```diff
@@ -42,8 +42,16 @@ export class Comment extends Component {
-		if ( nextProps.isBulkMode && ! this.props.isBulkMode ) {
-			this.setState( { isExpanded: false } );
+		if ( nextProps.isPostView ) {
+			if ( ! this.props.isPostView ) {
```
you can express this `if` as:

```js
if ( ! this.props.isPostView || this.props.isBulkMode !== nextProps.isBulkMode ) {
  this.setState( { isExpanded: false } );
}
```

or even shorter:
```js
componentWillReceiveProps( nextProps ) {
	const isExpanded = nextProps.isPostView
		? ! ( ! this.props.isPostView || this.props.isBulkMode !== nextProps.isBulkMode )
		: ! nextProps.isLoading && ! nextProps.isBulkMode;
	this.setState( { isExpanded } )
}
```

Copons replied:
I actually chose to be as clear as possible to avoid confusion on the logic.
If folks think a shorter approach is as clear or even more than my code, I'll happily update it! :)

Copons replied:
Actually, the ternary's first expression could end up setting `isExpanded = true`, which is not what we want.
My brain is melting on this thing and I'm gonna push a `let` solution. Sorry in advance! 😄

rodrigoi replied:
Maybe this component has too many `is` props. But don't stress too much about this. I just need to use the same approach on the Comment View, that's why I was revisiting this.

Copons commented:
Updates:

- The reply textarea is hidden by default.
- Added back the vertical margins.
- Changed the post link style.

Screenshots in PR description updated accordingly.

drw158 commented:
If you agree, let's do:

- 10px margin, not 16px
- Keep the post icon the same color as the link text (to be consistent with how the 'in reply to' icon looks. sorry, my bad!)
- Remove margins on mobile

And if we want to be more like the M3 mockups:

- Compact buttons for the actions

gwwar commented:
I'm really liking how this is shaping up ❤️  Do folks recall why we went with different behavior on the post specific view?

drw158 commented:
I don't recall, but it should probably have all expanded comments with no margins.

Copons commented:
> Do folks recall why we went with different behavior on the post specific view?

Because the threaded view basically only makes sense with collapsed comments.

Copons commented:
Another update I forgot before: the date is now _always_ relative if the comment is less than a month old.
I've moved the full date to the `title` attribute of the date permalink.

vindl commented:
I really like the new layout. :+1: Tested it out a bit and seems to be working as expected. There are a couple minor issues that I noticed:

- placeholders are still showing the old collapsed pattern, we should switch them to expanded cards with margins to match the new design:

![placeholders](https://user-images.githubusercontent.com/1182160/32671005-06b3aed8-c646-11e7-951a-bd46e88f88bc.png)

- After switching the pages/status the comments are first shown in collapsed view briefly, than they switch to all expanded without margins, and in the end the margins appear.

- Hitting the reply button will show the sliding avatar animation which might not make sense in this context when reply boxes are hidden by default.

Copons commented:
Thanks @vindl!
I noticed that there was something going on with the placeholders, but wasn't able to actually pinpoint what was wrong with them. 😄 

I think most of your findings will be fixed if we end up killing the collapsed view altogether.
So... let's kill it! 🔪

vindl commented on client/my-sites/comments/comment/index.jsx:
```diff
@@ -42,23 +42,35 @@ export class Comment extends Component {
+		const { isExpanded: wasExpanded, isReplyVisible: wasReplyVisible } = this.state;
+
+		let isExpanded = wasExpanded;
+		if ( isPostView && ( ! wasPostView || wasBulkMode !== isBulkMode ) ) {
```
This whole block for inferring `isExpanded` value looks confusing to me. Why do we need `wasPostView` to determine it, isn't it enough to know that we are currently in post view in order to show the collapsed version? (same goes for `wasBulkMode`). 

Also if we factor out `isLoading`, since I don't think it's needed anymore because we want to show expanded placeholders, it looks to me like all of this could be simplified to:

`const isExpanded = ! isPostView && ! isBulkMode;`

gwwar commented:
> Because the threaded view basically only makes sense with collapsed comments.

We went over this in p7jreA-1lw-p2 comments, no? 😄

gwwar commented:
I think this should be good to 🚢  after we match behavior in the post specific view, and the canary is happy. @drw158 did you have anything else you'd like changed?

Copons commented:
Big update, plenty of screens!

- Killed the collapsed view (1bdf500).
- Made the placeholders look as generic as possible.
- Removed card margins on small screens.
- Reworked the "Pending" label positioning (now it's `float: right`) to make it behave even when there is no "Post Title" (e.g. in Post View) and "In Reply To" line. Meaning: the first couple of comment text lines will wrap around the label.

Site View
<img width="720" alt="screen shot 2017-11-13 at 14 21 57" src="https://user-images.githubusercontent.com/2070010/32730456-0ee374f4-c87f-11e7-8047-5dfcfcebcac8.png">

Site View (small screen)
<img width="532" alt="screen shot 2017-11-13 at 14 23 40" src="https://user-images.githubusercontent.com/2070010/32730507-2fef024e-c87f-11e7-94c9-28fff6c34a0d.png">

Bulk Mode
<img width="717" alt="screen shot 2017-11-13 at 14 22 29" src="https://user-images.githubusercontent.com/2070010/32730519-385a2c2e-c87f-11e7-836b-79e7e4cdd482.png">

Reply Mode
<img width="720" alt="screen shot 2017-11-13 at 14 21 48" src="https://user-images.githubusercontent.com/2070010/32730482-2054c134-c87f-11e7-9e06-c0407d6ffec5.png">

Edit Mode
<img width="718" alt="screen shot 2017-11-13 at 14 22 08" src="https://user-images.githubusercontent.com/2070010/32730533-3f6d81aa-c87f-11e7-80f2-fefe516c18a5.png">

Post View
<img width="716" alt="screen shot 2017-11-13 at 14 24 45" src="https://user-images.githubusercontent.com/2070010/32730544-5bd37cf0-c87f-11e7-99d2-2f454fd16ab4.png">

Pending Comment
<img width="718" alt="screen shot 2017-11-13 at 14 21 30" src="https://user-images.githubusercontent.com/2070010/32730424-f6b961e0-c87e-11e7-98df-12b6a06d905c.png">

Placeholders
<img width="711" alt="screen shot 2017-11-13 at 14 34 41" src="https://user-images.githubusercontent.com/2070010/32730702-d3781b3a-c87f-11e7-93fd-b2b40ea6ea7a.png">

drw158 commented:
I really like it. Thank you for updating with screenshots!

I think we should

- Disable/remove avatar animation in reply box for now (meaning, just remove/hide the avatar). Right now, the user would only see it if they click outside of the field and then click inside again. Currently when you hit Reply, the avatar animates so fast, it's not helpful.
- We should probably remove the post title in bulk edit mode on the single post view?
- Now the the border on the header is gone, we need to decrease the margin directly below the header (needs to work on both regular and single post view). Notice how the space looks a bit big:

![screen shot 2017-11-13 at 10 12 08 am](https://user-images.githubusercontent.com/618551/32735751-315af4e0-c85b-11e7-84a6-5b545e1c8cc0.png)
![screen shot 2017-11-13 at 10 11 37 am](https://user-images.githubusercontent.com/618551/32735752-3168893e-c85b-11e7-91e7-b788ad7f67ea.png)

Not sure how to best do this though, since the first element in `comment__content` is not always the same.

vindl commented on client/my-sites/comments/comment/index.jsx:
```diff
@@ -37,60 +37,49 @@ export class Comment extends Component {
-			isExpanded: true,
-			isReplyVisible: ! isReplyVisible,
-		} ) );
+		this.setState( ( { isReplyVisible } ) => ( { isReplyVisible: ! isReplyVisible } ) );
```
Nitpick: we can simplify this further by passing an object instead of a function:

`this.setState( { isReplyVisible: ! isReplyVisible } );`

Copons replied:
Nope!
Here the reason: https://reactjs.org/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous 🙂

vindl replied:
Oh right :+1:

Copons commented:
As already mentioned, https://github.com/Automattic/wp-calypso/pull/19560#discussion_r150602908 fixed https://github.com/Automattic/wp-calypso/pull/19560#pullrequestreview-76170063 😄

Copons commented:
Forgot to mention, but @drw158 I took care of everything you mentioned in https://github.com/Automattic/wp-calypso/pull/19560#issuecomment-343972184

gwwar commented:
@Copons when clicking on reply, I see 'send' in the top right of the card:
![screen shot 2017-11-13 at 11 55 10 am](https://user-images.githubusercontent.com/1270189/32746144-9cbeffb6-c869-11e7-9f6e-a874ef9ee6ef.png)

kwight commented:
Cool! A couple of little things for further PRs:
* I saw the same thing with `SEND` that @gwwar mentioned.
* Have we lost the edit block for existing dotcom users? I can edit my username, but it looks like it's unsuccessful – maybe it's just the styling (the grayed-out name and URL) that we've lost.
* Smaller screens have some actions cut off.

![screen shot 2017-11-13 at 12 26 23 pm](https://user-images.githubusercontent.com/349751/32747548-f1395d26-c86d-11e7-9021-83ac5a7151f0.png)

Copons commented:
@gwwar @kwight The Send position got screwed up when I removed the reply gravatar and its transition. Good catch, I totally forgot about it! 😮 

The edit author: I've enabled the edit "self" because it just made sense. I didn't realize that it wasn't possible at API level. Disabled it again, for good. Hopefully I'll remember about it next time. 😄 

The actions: I've hidden their labels on `<480px`, but imho we should find a better design approach, in a follow up PR (cc @drw158).
